### PR TITLE
Remove linters from requirements-tests.txt

### PR DESCRIPTION
**Pull request by AlexWaygood**
These days, we do all our linting in CI via pre-commit; linters are only listed in `requirements-tests.txt` for local convenience, as pre-commit uses an isolated virtual environment for each linting tool it executes. But running `pre-commit run -a black` isn't really much more to type than `black`, and installing all these linters into the same environment is starting to cause issues. (See: the CI failures in https://github.com/python/typeshed/pull/12717.) Flake8-pyi has dropped support for Python 3.8 in its latest version, and other linters are likely to soon follow, but per the consensus in https://github.com/python/typeshed/issues/12716 we probably won't do so until the new year. If we're happy to do our linting via pre-commit locally as well as doing it via pre-commit in CI, that makes the situation caused by some tools dropping support for Python 3.8 quite a lot less problematic.

The only linter that I think we do need to keep in `requirements-tests.txt` is Ruff, because we run it as a subprocess from `create_baseline_stubs.py`, and we run it in that script with slightly different arguments to the ones we use in our pre-commit config file.

If we don't want to go this route, then we can just use environment markers in the `requirements-tests.txt` file to specify that the linting dependencies only need to be installed on Python 3.9+. We did this for a while after flake8 and flake8-pyi dropped support for Python 3.7 (until https://github.com/python/typeshed/commit/487e33112e98c07f246de6494a818ab50bf2e289), and it worked fine. The main disadvantages of that approach are (1) that it feels like a bit of a hack, and (2) we'll probably run into the same issue again when Python 3.9 goes EOL in around a year's time. This approach solves it for the long term.

### CODE REVIEW

**Review by srittau — APPROVED** [review]

**Comment by AlexWaygood** [comment]
I'll leave this open for a bit in case this severely breaks the workflow of any other maintainers :-)

**Comment by Avasam** [comment]
I believe uv supports https://peps.python.org/pep-0723/ , it might help go in that direction by supporting `uv run tests/runtests.py`. Although we might end up with duplicate specifications? idk, something to consider.

**Comment by AlexWaygood** [comment]
> I believe uv supports [peps.python.org/pep-0723](https://peps.python.org/pep-0723/) , it might help go in that direction by supporting `uv run tests/runtests.py`. Although we might end up with duplicate specifications? idk, something to consider.

yeah, I think it might be quite difficult to keep requirements pins synchronised between our various scripts (and keep the pins up-to-date with renovate) if we have different PEP-723 pins in each script. Call me old-fashioned, but I also just quite like to be able to run scripts with just `python` rather than `uv run` :-)

**Review by Avasam — APPROVED** [review]

File: requirements-tests.txt [hunk 1]
```diff
@@ -1,16 +1,9 @@
-# Type checkers and other linters that we test our stubs against. These should always
-# be pinned to a specific version to make failure reproducible. See also the
-# "tool.typeshed" section in pyproject.toml for additional type checkers.
-black==24.8.0            # must match .pre-commit-config.yaml
-flake8==7.1.1            # must match .pre-commit-config.yaml
-flake8-noqa==1.4.0       # must match .pre-commit-config.yaml
-flake8-pyi==24.6.0       # must match .pre-commit-config.yaml
```
Comment (Avasam) [inline requirements-tests.txt:7]: I just realized after approving, you can remove the `# must match requirements-tests.txt` text in `.pre-commit-config.yaml`

**Pull request merged**